### PR TITLE
Fanctl2 - state machine, fan start, random phase shifting

### PR DESCRIPTION
**Pull request by XPila**
BFW-918

### CODE REVIEW

File: src/common/fanctl.h [hunk 1]
```diff
@@ -52,6 +50,13 @@ typedef struct _fanctl_tach_t {
 
 // class for software pwm control with phase-shifting
 class CFanCtlPWM : private fanctl_pwm_t {
+public:
+    enum PhaseShiftMode {
```
Comment (DRracer) [inline src/common/fanctl.h:54]: probably `enum PhaseShiftMode : uint8_t {` , there are only few items, so we may save some RAM

File: src/common/fanctl.h [hunk 2]
```diff
@@ -65,9 +70,11 @@ class CFanCtlPWM : private fanctl_pwm_t {
     inline uint8_t get_min_PWM() { return min_value; }
     inline uint8_t get_max_PWM() { return max_value; }
     inline uint8_t get_PWM() { return pwm; }
+    inline PhaseShiftMode get_PhaseShiftMode() { return (PhaseShiftMode)pha_mode; }
```
Comment (DRracer) [inline src/common/fanctl.h:73]: const

File: src/common/fanctl.h [hunk 3]
```diff
@@ -65,9 +70,11 @@ class CFanCtlPWM : private fanctl_pwm_t {
     inline uint8_t get_min_PWM() { return min_value; }
     inline uint8_t get_max_PWM() { return max_value; }
     inline uint8_t get_PWM() { return pwm; }
```
Comment (DRracer) [inline src/common/fanctl.h:72]: const ... and the other methods as well...

File: src/common/fanctl.cpp [hunk 4]
```diff
@@ -48,22 +49,31 @@ int8_t CFanCtlPWM::tick() {
...
+                if (pha >= pha_max) {
+                    pha_stp = -pha_stp;
+                    pha = pha_max;
+                } else if (pha < 0) {
+                    pha_stp = -pha_stp;
+                    pha = 0;
+                }
+                break;
+            case random:
+                pha = pha_max * ((float)rand() / RAND_MAX);
```
Comment (DRracer) [inline src/common/fanctl.cpp:74]: the float probably won't be too healthy for the 8bit, but is ok here

Comment (XPila) [reply src/common/fanctl.cpp:75]: you are right, for 8bit we cannot use float, i will try to find common solution with fixpoint aritmetic or somthing similar

File: src/common/fanctl.h (hunk 1 above)
Comment (XPila) [reply src/common/fanctl.h:55]: fixed. internaly it is stored in pha_mode (uint8_t).

File: src/common/fanctl.h (hunk 2 above)
Comment (XPila) [reply src/common/fanctl.h]: fixed

File: src/common/fanctl.h (hunk 3 above)
Comment (XPila) [reply src/common/fanctl.h]: fixed

File: src/common/fanctl.h (hunk 1 above)
Comment (DRracer) [reply src/common/fanctl.h]: almost, but ok...

**Review by RadekVana — APPROVED** [review]

**Pull request merged**